sound/soc/sof/intel/hda-mlink.c

```diff
@@ -14,28 +14,235 @@
+			  void __iomem *ml_addr, int link_idx)
+{
+	struct hdac_ext_link *hlink = &h2link->hext_link;
+	u32 base_offset;
+
+	hlink->lcaps  = readl(ml_addr + AZX_REG_ML_LCAP);
+
+	h2link->alt = FIELD_GET(AZX_ML_HDA_LCAP_ALT, hlink->lcaps);
+
+	/* handle alternate extensions */
+	if (!h2link->alt) {
+		/*
+		 * LSDIID is initialized by hardware for HDaudio link,
+		 * it needs to be setup by software for alternate links
+		 */
+		hlink->lsdiid = readw(ml_addr + AZX_REG_ML_LSDIID);
+
+		dev_dbg(dev, "Link %d: HDAudio - lsdiid=%d\n",
+			link_idx, hlink->lsdiid);
+	} else {
+		h2link->intc = FIELD_GET(AZX_ML_HDA_LCAP_INTC, hlink->lcaps);
+		h2link->ofls = FIELD_GET(AZX_ML_HDA_LCAP_OFLS, hlink->lcaps);
+		h2link->lss = FIELD_GET(AZX_ML_HDA_LCAP_LSS, hlink->lcaps);
+
+		/* read slcount (increment due to zero-based hardware representation */
+		h2link->slcount = FIELD_GET(AZX_ML_HDA_LCAP_SLCOUNT, hlink->lcaps) + 1;
+		dev_dbg(dev, "Link %d: HDAudio extended - sublink count %d\n",
+			link_idx, h2link->slcount);
+
+		/* find IP ID and offsets */
+		h2link->leptr = readl(hlink->ml_addr + AZX_REG_ML_LEPTR);
+
+		h2link->elid = FIELD_GET(AZX_REG_ML_LEPTR_ID, h2link->leptr);
+
+		base_offset = FIELD_GET(AZX_REG_ML_LEPTR_PTR, h2link->leptr);
+		h2link->base_ptr = hlink->ml_addr + base_offset;
+
+		switch (h2link->elid) {
+		case AZX_REG_ML_LEPTR_ID_SDW:
+			h2link->shim_offset = AZX_REG_SDW_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_SDW_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_SDW_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - SoundWire alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_DMIC:
+			h2link->shim_offset = AZX_REG_INTEL_DMIC_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_DMIC_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_DMIC_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL DMIC alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_SSP:
+			h2link->shim_offset = AZX_REG_INTEL_SSP_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_SSP_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_SSP_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL SSP alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_UAOL:
+			h2link->shim_offset = AZX_REG_INTEL_UAOL_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_UAOL_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_UAOL_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL UAOL alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		default:
+			dev_err(dev, "Link %d: HDAudio extended - Unsupported alternate link, leptr.id=%#02x value\n",
+				link_idx, h2link->elid);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+/* END HDAML section */
+
+static int hda_ml_alloc_h2link(struct hdac_bus *bus, int i)
```

Comment:
can we call this second arg index as well, would be cleaner to read

Reply:
done